src/python/pants/backend/python/lint/yapf/rules_integration_test.py

```diff
@@ -0,0 +1,261 @@
+    lint_results = rule_runner.request(LintResults, [YapfRequest(field_sets)])
+    input_sources = rule_runner.request(
+        SourceFiles,
+        [
+            SourceFilesRequest(field_set.sources for field_set in field_sets),
+        ],
+    )
+    fmt_result = rule_runner.request(
+        FmtResult,
+        [
+            YapfRequest(field_sets, prior_formatter_result=input_sources.snapshot),
+        ],
+    )
+    return lint_results.results, fmt_result
+
+
+def get_digest(rule_runner: RuleRunner, source_files: dict[str, str]) -> Digest:
+    files = [FileContent(path, content.encode()) for path, content in source_files.items()]
+    return rule_runner.request(Digest, [CreateDigest(files)])
+
+
+def test_passing_source(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files({"f.py": GOOD_FILE, "BUILD": "python_library(name='t')"})
+    tgt = rule_runner.get_target(Address("", target_name="t", relative_file_path="f.py"))
+    lint_results, fmt_result = run_yapf(rule_runner, [tgt])
+    assert len(lint_results) == 1
+    assert lint_results[0].exit_code == 0
+    assert lint_results[0].stderr == ""
+    assert fmt_result.output == get_digest(rule_runner, {"f.py": GOOD_FILE})
+    assert fmt_result.did_change is False
+
+
+def test_failing_source(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files({"f.py": BAD_FILE, "BUILD": "python_library(name='t')"})
+    tgt = rule_runner.get_target(Address("", target_name="t", relative_file_path="f.py"))
+    lint_results, fmt_result = run_yapf(rule_runner, [tgt])
+    assert len(lint_results) == 1
+    assert lint_results[0].exit_code == 1
+    assert all(msg in lint_results[0].stdout for msg in ("reformatted", "original"))
+    assert fmt_result.skipped is False
+    assert fmt_result.output == get_digest(rule_runner, {"f.py": FIXED_BAD_FILE})
+    assert fmt_result.did_change is True
+
+
+def test_multiple_targets(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {"good.py": GOOD_FILE, "bad.py": BAD_FILE, "BUILD": "python_library(name='t')"}
+    )
+    tgts = [
+        rule_runner.get_target(Address("", target_name="t", relative_file_path="good.py")),
+        rule_runner.get_target(Address("", target_name="t", relative_file_path="bad.py")),
+    ]
+    lint_results, fmt_result = run_yapf(rule_runner, tgts)
+    assert len(lint_results) == 1
+    assert lint_results[0].exit_code == 1
+    assert all(msg in lint_results[0].stdout for msg in ("reformatted", "original", "bad.py"))
+    assert "good.py" not in lint_results[0].stdout
+    assert fmt_result.output == get_digest(
+        rule_runner, {"good.py": GOOD_FILE, "bad.py": FIXED_BAD_FILE}
+    )
+    assert fmt_result.did_change is True
+
+
+@pytest.mark.parametrize(
+    "path,section,extra_args",
+    (
+        (".style.yapf", "style", []),
+        ("setup.cfg", "yapf", []),
+        ("pyproject.toml", "tool.yapf", []),
+        ("custom.style", "style", ["--yapf-config=custom.style"]),
```

Comment:
Nice! This is really robust, altho imo we may want to only test one of the auto-discovered config formats + the `custom.style`. There's some performance impact to running these `RuleRunner` tests, and it's always a trade-off between robustness vs. CI taking more time. 

We already have pretty robust tests for the `ConfigFiles` mechanism, so I think it's probably fine to only have one smoke screen test that config autodiscovery is wired correctly. `custom.style` is great to keep because it requires custom logic in setting up `--argv`.

Reply:
Thanks for mentioning the CI time. I've noticed that it's already taking a lot of time to run all the tests, so we definitely don't want to make it even slower. I've followed your suggestion and kept only a single "standard" style file test and one custom style file.